### PR TITLE
Fixes: missing functionality for svg <use> elements 

### LINKED ISSUE #652: missing functionality for svg <use> elements
With the `ImageSVG` functionality,
[image](https://www.w3.org/Graphics/SVG/)
is not displayed correctly.
image

**location of missing functionality** `startSVGElementHandler` and `endSVGElementHandler`.

**behavior in version 6.6.5**
- objects outside `<defs>` elements can not be referenced with `<use>` since they are not stored in array `$svgdefs`.
- `$svgdefs` does not contain child elements with id attributes set.

**additional info**
- currently the structure does not allow for elements to be stored as child element and independently being referenced. 
- a `<use>` element within a recursive call from a `<use>` can cause issues when addressing the points above.

### PR DESCRIPTION
**Pull request by ghost**
Fixes:  #652

### CODE REVIEW

File: tcpdf.php [hunk 1]
```diff
@@ -23944,21 +23958,33 @@ protected function startSVGElementHandler($parser, $name, $attribs, $ctm=array()
 			$this->svgclippaths[$this->svgclipid][] = array('name' => $name, 'attribs' => $attribs, 'tm' => $this->svgcliptm[$this->svgclipid]);
 			return;
 		}
-		if ($this->svgdefsmode AND !in_array($name, array('clipPath', 'linearGradient', 'radialGradient', 'stop'))) {
+		if (!in_array($name, array('clipPath', 'linearGradient', 'radialGradient', 'stop')) AND $this->svgusedepth == 0) {
```
Comment (williamdes) [inline tcpdf.php:23961]:
```suggestion
		if (!in_array($name, array('clipPath', 'linearGradient', 'radialGradient', 'stop')) AND $this->svgusedepth === 0) {
```

File: tcpdf.php [hunk 2]
```diff
@@ -1641,6 +1641,20 @@ class TCPDF {
 	 */
 	protected $svgdefs = array();
 
+    /**
```
Comment (williamdes) [inline tcpdf.php:1644]: I am afraid this patch is not very minimal and so will have less changes being accepted into tcpdf

**Review by williamdes — APPROVED** [review]
Thank for making this patch !

File: tcpdf.php (hunk 2 above)
Comment (k-nicole) [reply tcpdf.php:1644]: This library is part of the moodle core. We at TU Graz are relying on it to generate certificates, which among other contents contain logos of our partners (SVGs). Its crucial for us that they are correct on the pdf and currently this is not the case. Hence our contribution to this library.
Is there anything that can be done to increase chances? I will change up the contribution in any way necessary to provide this functionality. (:

Comment (williamdes) [reply tcpdf.php:1644]: First, thank you for contributing to the libraries you use !

Second, I am here as an advisor and try to monitor issues and pull-requests on my free time
Currently the library owner and maintainer remains mysterious in choices that are made. But it's mostly the rule of more the change is okay and small more it has chances to get on a release one day.

So from time to time I build an issue with a summary of PRs: https://github.com/tecnickcom/TCPDF/issues/634
Maybe it helps, I never know, nobody knows
So it's a wait and see, but you need to get it slim and clean and right. Or it will probably get refused and you are back in a 6-8 months loop to get feedback on the new changes

This recent PR changed the library maintenance official statement, and that's a relief: #589 

In summary, I think this patch will appear too large. Is there other ways to change less code ?

**Comment by BsAtHome** [comment]
Hi,
There are still a few problems with this patch applied. The `<use>` element is more complicated than implemented by this patch. I have a problem file that shows the issue(s) (a cooked down test version from the original shown below).
There are three clipPath scenarios that I tested.
* First is `<use>` in the `<defs><clipPath>` section with a forward reference to a path later on in the document. This scenario fails completely and no clipPath is applied.
* Second is a copied version of the document's `<path>` in the `<defs><clipPath>` section, but it retained the original transformation attribute from the source. This scenario failed to apply the transformation.
* Third is a working version, where the path is normalized.

The used svg (inkscape displays all three version correct):
`<svg id="testUseClippath" width="100%" height="100%" viewBox="0 0 50 30" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">`
`  <defs>`
`    <clipPath id="usepath"><use xlink:href="#realpath" /></clipPath>`
`<!--    <clipPath id="usepath"><path id="realpath" d="M 14,-1 C 10,1 6,9 0,9 C -6,9 -12,1 -14,-1 C -11,-2 -6,-9 0,-9 C 6,-9 11,-2 14,-1 Z" transform="translate(25,15)" /></clipPath> -->`
`<!--    <clipPath id="usepath"><path d="M 39,14 C 35,16 31,24 25,24 C 19,24 13,16 11,14 C 14,13 19,6 25,6 C 31,6 36,13 39,14 Z" /></clipPath> -->`
`  </defs>`
`  <g>`
`    <rect x="0.1" y="0.1" width="49.8" height="29.8" style="fill:none;stroke:#cc0;stroke-width:0.2" />`
`    <g clip-path="url(#usepath)">`
`      <rect x="0" y="0" width="50" height="30" style="fill:#222;stroke:none" />`
`      <ellipse cx="25" cy="25" rx="15" ry="10" style="fill:#800f08;stroke:none" />`
`    </g>`
`    <path id="realpath" d="M 14,-1 C 10,1 6,9 0,9 C -6,9 -12,1 -14,-1 C -11,-2 -6,-9 0,-9 C 6,-9 11,-2 14,-1 Z" transform="translate(25,15)" style="fill:none;stroke:black;stroke-width:1" />`
`  </g>`
`</svg>`

**Comment by nicolaasuni** [comment]
Please try to resolve the conflicts.
